### PR TITLE
Channel Module Hardening and Feature Enhancement

### PR DESCRIPTION
**Pull request by jpirnay**
This PR improves the channel.py module by addressing critical issues, adding new features, and implementing comprehensive testing. The changes enhance thread safety, memory management, error resilience, and maintainability while preserving full backward compatibility.

### 🔧 Critical Bug Fixes
- Fixed E731 linting issue: Converted lambda assignment to proper def function
- Resolved BBCode parsing: Extended regex to handle closing tags ([/red], [/bold], etc.)
- Thread safety overhaul: Replaced unsafe list-based queue with thread-safe queue.Queue
- Error isolation: Added try/catch blocks around watcher calls to prevent one broken watcher from crashing the entire channel system
### 🚀 New Features
- Weak reference support: Added optional weak references for watchers to prevent memory leaks
- Dynamic buffer management: Added resize_buffer() method for runtime buffer size changes
- Enhanced documentation: Comprehensive class and method docstrings with usage examples
### 🧪 Testing & Quality
- Comprehensive test suite: Added 26 unit tests covering all functionality
- Thread safety validation: Tests for concurrent operations
- Error handling verification: Tests ensure watcher failures don't break the system
- Memory leak prevention: Tests validate weak reference cleanup

## Summary by Sourcery

Harden the Channel module by fixing critical parsing and threading issues, adding robust error isolation, memory-safe weak references, dynamic buffer management, and enriched documentation, all backed by full test coverage

New Features:
- Support optional weak references for watcher functions
- Allow dynamic resizing of channel buffer at runtime

Bug Fixes:
- Extend BBCode regex to handle missing closing tags and correct ANSI code for [/italic]
- Replace unsafe list-based queue with thread-safe Queue in threaded channels
- Add exception handling around watcher invocations to prevent a failing watcher from crashing the system
- Prevent duplicate watcher registrations and fix unwatch removal of weak references

Enhancements:
- Integrate a SimpleLogger class for structured logging
- Refactor watcher dispatch with a unified _call_watcher helper
- Enhance class and method docstrings with usage examples

Tests:
- Add a comprehensive unit test suite covering core functionality, threading, error handling, BBCode conversion, buffer management, and weak reference behavior

### CODE REVIEW

**Comment by jpirnay** [comment]
@sourcery-ai review

File: test/test_channel.py [hunk 1]
```diff
@@ -0,0 +1,382 @@
...
+class TestChannelThreading(unittest.TestCase):
+    """Test threaded channel functionality."""
+
+    def setUp(self):
+        self.channel = Channel("threaded")
+
+    def tearDown(self):
+        if hasattr(self.channel, 'threaded') and self.channel.threaded:
+            self.channel.threaded = False
+
```
Comment (SourceryAI) [inline test/test_channel.py:431]: **suggestion (testing):** Threaded channel tests are placeholders and do not validate actual threading behavior.

Please add integration tests that start the channel in threaded mode, send messages from multiple threads, and verify correct, thread-safe message delivery.

File: test/test_channel.py [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+    def test_buffering(self):
+        """Test message buffering functionality."""
+        channel = Channel("buffered", buffer_size=3)
+
+        # Send messages
+        channel("msg1", indent=False)
+        channel("msg2", indent=False)
+        channel("msg3", indent=False)
+        channel("msg4", indent=False)  # Should push out msg1
+
```
Comment (SourceryAI) [inline test/test_channel.py:114]: **suggestion (testing):** Buffering tests do not cover disabling and resizing the buffer at runtime.

Please add tests for the resize_buffer method to cover shrinking, growing, and disabling the buffer, and verify the resulting buffer contents and behavior.

File: test/test_channel.py [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
+        channel = Channel("ansi", ansi=True)
+
+        # Test basic colors
+        result = channel.bbcode_to_ansi("[red]red text[/red]")
+        self.assertIn("\033[31m", result)
+        self.assertIn("\033[0m", result)
+
+        # Test raw passthrough
+        result = channel.bbcode_to_ansi("[raw][red]raw bbcode[/raw]")
+        self.assertIn("[red]", result)
```
Comment (SourceryAI) [inline test/test_channel.py]: **suggestion (testing):** BBCode to ANSI conversion tests do not cover all supported tags and edge cases.

Please add parameterized tests for all BBCode tags, including closing, nested, and malformed cases, to improve coverage and robustness.

```suggestion
    def test_bbcode_to_ansi_conversion(self):
        """Test BBCode to ANSI conversion, including all supported tags and edge cases."""
        channel = Channel("ansi", ansi=True)

        bbcode_cases = [
            # Basic colors
            ("[red]red text[/red]", "\033[31m", "\033[0m"),
            ("[green]green text[/green]", "\033[32m", "\033[0m"),
            ("[blue]blue text[/blue]", "\033[34m", "\033[0m"),
            # Styles
            ("[bold]bold text[/bold]", "\033[1m", "\033[0m"),
            ("[italic]italic text[/italic]", "\033[3m", "\033[0m"),
            # Nested tags
            ("[red][bold]red bold[/bold][/red]", "\033[31m", "\033[1m"),
            # Malformed tags
            ("[red]no closing", "\033[31m", None),
            ("[red][green]nested no close[/red]", "\033[31m", "\033[32m"),
            ("[red][bold]missing end", "\033[31m", "\033[1m"),
            # Closing tags only
            ("[/red]", None, None),
            # Raw passthrough
            ("[raw][red]raw bbcode[/raw]", "[red]", None),
        ]

        for bbcode, expected_ansi, expected_ansi2 in bbcode_cases:
            with self.subTest(bbcode=bbcode):
                result = channel.bbcode_to_ansi(bbcode)
                if expected_ansi:
                    self.assertIn(expected_ansi, result)
                if expected_ansi2:
                    self.assertIn(expected_ansi2, result)
                if "[raw]" in bbcode:
                    self.assertIn("[red]", result)
```

File: test/test_channel.py [hunk 4]
```diff
@@ -0,0 +1,382 @@
...
+
+    def test_watcher_exception_isolation(self):
+        """Test that one watcher exception doesn't break others."""
+        messages = []
+
+        def good_collector(msg):
+            messages.append(f"good: {msg}")
+
+        def bad_collector(msg):
+            raise Exception("Watcher failed")
```
Comment (SourceryAI) [inline test/test_channel.py:575]: **issue (code-quality):** Raise a specific error instead of the general `Exception` or `BaseException` ([`raise-specific-error`](https://docs.sourcery.ai/Reference/Rules-and-In-Line-Suggestions/Python/Default-Rules/raise-specific-error))

<details><summary>Explanation</summary>If a piece of code raises a specific exception type
rather than the generic
[`BaseException`](https://docs.python.org/3/library/exceptions.html#BaseException)
or [`Exception`](https://docs.python.org/3/library/exceptions.html#Exception),
the calling code can:

- get more information about what type of error it is
- define specific exception handling for it

This way, callers of the code can handle the error appropriately.

How can you solve this?

- Use one of the [built-in exceptions](https://docs.python.org/3/library/exceptions.html) of the standard library.
- [Define your own error class](https://docs.python.org/3/tutorial/errors.html#tut-userexceptions) that subclasses `Exception`.

So instead of having code raising `Exception` or `BaseException` like

```python
if incorrect_input(value):
    raise Exception("The input is incorrect")
```

you can have code raising a specific error like

```python
if incorrect_input(value):
    raise ValueError("The input is incorrect")
```

or

```python
class IncorrectInputError(Exception):
    pass


if incorrect_input(value):
    raise IncorrectInputError("The input is incorrect")
```
</details>

**Comment by jpirnay** [comment]
@sourcery-ai resolve

**Comment by jpirnay** [comment]
@sourcery-ai review

**Review by jpirnay — COMMENTED** [review]
Needs more work
